packages/client/src/utils/buttonActions.js

```diff
@@ -0,0 +1,45 @@
+import { enrichDataBinding } from "./enrichDataBinding"
+import { routeStore } from "../store"
+import { saveRow, deleteRow } from "../api"
+
+const saveRowHandler = async (action, context) => {
+  let draft = context[`${action.parameters.contextPath}_draft`]
+  if (action.parameters.fields) {
+    Object.entries(action.parameters.fields).forEach(([key, entry]) => {
+      draft[key] = enrichDataBinding(entry.value, context)
+    })
+  }
+  await saveRow(draft)
+}
+
+const deleteRowHandler = async (action, context) => {
+  const { tableId, revId, rowId } = action.parameters
+  await deleteRow({
+    tableId: enrichDataBinding(tableId, context),
+    rowId: enrichDataBinding(rowId, context),
+    revId: enrichDataBinding(revId, context),
+  })
+}
+
+const navigationHandler = action => {
+  routeStore.actions.navigate(action.parameters.url)
+}
+
+const handlerMap = {
+  ["Save Row"]: saveRowHandler,
+  ["Delete Row"]: deleteRowHandler,
+  ["Navigate To"]: navigationHandler,
+}
+
+/**
+ * Parses an array of actions and returns a function which will execute the
+ * actions in the current context.
+ */
+export const enrichButtonActions = (actions, context) => {
+  const handlers = actions.map(def => handlerMap[def["##eventHandlerType"]])
+  return async () => {
+    for (let i = 0; i < handlers.length; i++) {
+      await handlers[i](actions[i], context)
```

Comment:
NAB: can run these handlers concurrently with `Promise.all`

Reply:
Good catch. I think they should probably run consecutively though due to the nature of the promises themselves - like make sure a row is saved before navigating away from the page, or `navigate` > `doSomething` > `navigate` (although I can't see why someone would do that...). What do you reckon?

Reply:
Good point! That makes sense - since you can have chains of actions it might cause strange behaviour if we run all the "steps" simultaneously.